### PR TITLE
Add atmos leakage system + Diffuser fans

### PR DESCRIPTION
**Pull request by SlamBamActionman**
## About the PR

This PR adds functionality for **atmos leakage**; tiles where atmos transfer happens at a gradual rate, rather than equalizing the air mix as fast as possible. As a way to introduce this functionality the PR adds a new item: the **Diffuser Fan.** 

## Why / Balance

Atmos is currently very much an all-or-nothing approach, attempting to equalize the pressure of a space as fast as possible; while this works just fine for most scenarios, it's also a bit inflexible. The atmos leakage functionality introduces more granularity to the atmos system by setting certain tiles to only let through a ratio of the original transfer amount. The transfer reduction is set as a ratio of the max possible transfer; a 50% reduction of a 10 mole transfer would only allow 5 moles to pass through, while a 1 mole transfer would only allow 0.5 moles.

In the future this functionality could be used for something like cracked windows, though for this PR it is implemented via the new Diffuser Fan.

The Diffuser Fan is a mapper-only item that acts similar to a tiny fan, fitting inside airlocks and allowing crew to traverse over it.
The intended purpose is to put them inside evac shuttles, placing them inside internal airlocks and firelocks inside the shuttle. This would have the following consequences:

> Shuttlebombing lethality would be reduced for *non-target* crew. It is trivially easy to accidentally space non-bombed areas of the shuttle by opening air/firelocks, either to escape the spacing or intending to save those in the spaced areas. Introducing diffuser fans would *not* make shuttle bombing less lethal for the target of the bomb or those in the immediate viscinity, just reduce casualties among bystanders in other parts of the shuttle. Since atmos transfer does still happen it does not give a free safe space like tiny fans do, just more flexibility for crew to maneouver out of the spaced area.

The diffuser fan is given a 50% atmos transfer reduction. This value needs to be tested further to see if it's appropriate for its intended purpose. Another possible implementation of the diffuser fan would be in the Reclaimer airlocks, which currently have no fans at all.

## Technical details

This PR introduces the `AirLeakage` component and system, which are used to modify the new `TransferRatio` property for `TileAtmosphere`. 

`TransferRatio` is used to determine the ratio for atmos is transferred across tiles, multiplying the lowest `TransferRatio` property of two tiles exchanging atmos with the original atmos transfer amount. 

Much of the atmospherics code is intended to optimize atmos equalization across a region. Since tiles with a modified `TransferRatio` are intended to equalize slower by definition, these tiles are excluded from this optimization. The impact of this on performance has yet to be evaluated, though it's assumed that this would introduce about as much load as an entity breathing. 

The `AirLeakage` implementation is based on the structure of the `Airtight` component and system.

## Media

[Youtube video](https://youtu.be/wCWmCD_-cqg)

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

**Changelog**

No CL, this PR doesn't introduce any player-facing changes.

### CODE REVIEW

**Comment by SlamBamActionman** [comment]
I've recorded a video using a Diffuser Fan value of 0.5: https://youtu.be/wCWmCD_-cqg

To sum up the differences:
- Normal time to reach barotrauma pressure was 11 seconds, 16 seconds to fully space. 
- Leakage time to reach barotrauma pressure was 21 seconds, 60 seconds to fully space.

The values could of course be adjusted, but this at least shows the concept.

**Comment by Partmedia** [comment]
Thank you for your PR, and sorry for the delay in reviews.

Conceptually, I think this could be okay simulating something like cracked windows and narrow gaps.

The part where I'm having difficulty seeing is the "diffuser fan".

> Atmos is currently very much an all-or-nothing approach, attempting to equalize the pressure of a space as fast as possible

Spacing is actually already quite slow right now. Slowing down atmos more would likely make less intuitive sense and be more difficult to troubleshoot.

> The intended purpose is to put them inside evac shuttles, placing them inside internal airlocks and firelocks inside the shuttle.

This only sounds marginally better than filling the shuttle with tiny fans.

> Shuttlebombing lethality would be reduced for non-target crew. It is trivially easy to accidentally space non-bombed areas of the shuttle by opening air/firelocks, either to escape the spacing or intending to save those in the spaced areas.

If shuttle bombing is the problem, why not focus on mechanics that reduce shuttle bombing instead of changing atmos?

> Another possible implementation of the diffuser fan would be in the Reclaimer airlocks, which currently have no fans at all.

There are interconnected airlocks, so why not use those? [Tiny fans are a crutch](https://discord.com/channels/310555209753690112/311537926376783886/1040711542234173512).

In summary, I think the system could be useful in the future for modeling things like cracked windows or narrow gaps, but I am going to need more convincing that diffuser fans are needed and good for the game. I'll make a pass through the code once we have some conceptual agreement here.

**Comment by SlamBamActionman** [comment]
Thanks for the response! I agree it's a very impactful mechanic, _especially_ the diffuser fan, so I'll be happy to elaborate:

The source of this mechanic originates from discussions regarding shuttle bombing and its lethality. There's a vocal part of the community that strongly objects to shuttle bombings even existing as a traitor strategy, and another part that equally strongly advocates for the chaos it causes. Maintainers have repeatedly reaffirmed that shuttle bombings are a desired mechanic due to the high-intensity gameplay it provides at the end of the round, but that doesn't change the dislike some people have for the strategy. 

So I thought I'd interrogate _why_ shuttle bombings are seen as cheap and unfun. There's a strong case to be said that it's far too reliable (you can count on your target being there, and if your target isn't on the shuttle you'll succeed with the maroon part of the objective anyways) which makes it common instead of novel. But someone running an e-sword or holoparasite on evac won't be nearly as derided in OOC; the main difference between that and bombings is the number of civilian casualties.

A bomb going off on the station isn't nearly as lethal to passerbys since it's easy to retreat away and, more importantly, any area adjacent to the spaced room can be easily kept safe by opening airlocks to other rooms. Evac does not have this benefit. Once a room is spaced it's incredibly lethal to open up to any adjacent rooms. This makes it hard to escape a spaced area, retrieve corpses, ~~walk in to repair (can't even do that)~~ and one errant borg accidentally passing by an airlock can spell doom for the entirety of medbay or Command.

Death is a core part of SS14 but I think it's fair to say that shuttle bombing has an incredibly high mortality rate for non-targets. Should it be dangerous? Yes! But since the traitor's target is already going to be in the middle of the bomb's blast I think it's reasonable to discuss ways to reduce bystander lethality.

> > The intended purpose is to put them inside evac shuttles, placing them inside internal airlocks and firelocks inside the shuttle.
> 
> This only sounds marginally better than filling the shuttle with tiny fans.

Tiny fans would be the first idea on how to solve the issues above, but it falls flat when you examine it closer. We can already observe what happens with tiny fans: they're barotrauma/airloss-safe tiles, so people just stack up multiple on them which a) looks really janky and b) isn't particulary interesting gameplay.

Diffuser fans are meant to provide an inbetween. They allow for safe(ish) traversal from a spaced to an unspaced area but are not abusable as a safe tile. Since it's just a normal tile with slower transfer it will always be at less pressure than the safe tile, and the longer you keep it open the more it and the safe area will lose pressure.

It's just a matter of finding the right value. Low enough that people can traverse the tile repeatedly, but high enough that holding the airlock open is undesirable. 

> > Atmos is currently very much an all-or-nothing approach, attempting to equalize the pressure of a space as fast as possible
> 
> Spacing is actually already quite slow right now. 

It's true that spacing isn't super fast; with the recent changes to barotrauma damage thresholds the time to start taking damage has been lowered further. I'd argue that in a standard shuttle the atmos leakage still happens fast enough to cause the issues described above. 

There's an additional little sneaky thing with the atmos system that can make it faster than the numbers make it out to be. Spacing slowdown is only applied upon calculations involving space tiles (i.e. tiles open to space). This means that if you're crowbaring into a spaced room that doesn't have a direct line to space the two rooms mix atmosphere instantly. On one hand this can be good to restore air to a previously dangerous area, but considering air is in short supply on a shuttle...

> Slowing down atmos more would likely make less intuitive sense and be more difficult to troubleshoot. 

I disagree that this introduces any heavy complexity, especially compared to how complex the system already is. The leakage system uses all the normal atmos calculations, it just adds a ratio on-top.
Since the diffuser fans are a map-only item, they should really only be utilized in controlled areas, i.e. evac shuttle, where troubleshooting would be fairly localized.

> > Shuttlebombing lethality would be reduced for non-target crew. It is trivially easy to accidentally space non-bombed areas of the shuttle by opening air/firelocks, either to escape the spacing or intending to save those in the spaced areas.
> 
> If shuttle bombing is the problem, why not focus on mechanics that reduce shuttle bombing instead of changing atmos?

Based on Discord discussions, maintainers have been hesitant to nerf shuttle bombing directly. Partially this has to do with most suggestions being kinda jank, and partially because shuttle bombing is meant to be an intended and viable mechanic. But the part that makes shuttle bombing so contentious is also the part that is hardest to balance; how do you tweak the lethality of a shuttlebomb when spacing is the real danger to most of the crew? Diffuser fans would be one such way.

I'm also of the opinion that diffuser fans would indirectly discourage shuttle bombing. You can no longer just chuck in a supermatter bomb and call it a day, since diffuser fans would allow crew and medical to much easier retrieve the body while also having less civilian casualties to deal with.

> > Another possible implementation of the diffuser fan would be in the Reclaimer airlocks, which currently have no fans at all.
> 
> There are interconnected airlocks, so why not use those? [Tiny fans are a crutch](https://discord.com/channels/310555209753690112/311537926376783886/1040711542234173512).

I was just assuming that it was somehow an intended feature to not have airlock cycling on the Reclaimer. I agree that the tiny fans are a crutch, they feel very game-y in the way they provide infinite atmos.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
I don't really know whether the changes to LINDA are a good idea but if you're adding new atmos code then PLEASE document it.

**Comment by SlamBamActionman** [comment]
> I don't really know whether the changes to LINDA are a good idea but if you're adding new atmos code then PLEASE document it.

Added quotes, though due to the code mostly being multiplying with a ratio it seems a bit excessive in places. Mostly added clarification that the value is going to be 1.0 in most cases.

The LINDA changes are necessary as otherwise the system believes tiles have been equalized when that isn't the case. This would result in cases where only a single tick of atmos calculation is performed and then nothing else, causing the leakage tile to effectively "stop leaking".

**Comment by Emisse** [comment]
Closed due to feature freeze May 10th-June 14th. Comment to have it reopen after this.
